File: ts/web3js/src/instructions/eata/delegateEphemeralAtaPermission.ts
import {
  PublicKey,
  TransactionInstruction,
  SystemProgram,
  AccountMeta,
} from "@solana/web3.js";
import {
  EATA_PROGRAM_ID,
  DELEGATION_PROGRAM_ID,
  PERMISSION_PROGRAM_ID,
} from "../../constants";
import {
  ephemeralAtaPdaWithBumpFromOwnerAndMint,
  permissionPdaFromAccount,
  delegateBufferPdaFromDelegatedAccountAndOwnerProgram,
  delegationRecordPdaFromDelegatedAccount,
  delegationMetadataPdaFromDelegatedAccount,
} from "../../pda";

export interface DelegateEphemeralAtaPermissionAccounts {
  payer: PublicKey;
  user: PublicKey;
  mint: PublicKey;
  validator: PublicKey;
}

export function createDelegateEphemeralAtaPermissionInstruction(
  accounts: DelegateEphemeralAtaPermissionAccounts,
): TransactionInstruction {
  const [ephemeralAta, bump] = ephemeralAtaPdaWithBumpFromOwnerAndMint(
    accounts.user,
    accounts.mint,
  );
  const permission = permissionPdaFromAccount(ephemeralAta);
  const buffer = delegateBufferPdaFromDelegatedAccountAndOwnerProgram(
    permission,
    PERMISSION_PROGRAM_ID,
  );
  const record = delegationRecordPdaFromDelegatedAccount(permission);
  const metadata = delegationMetadataPdaFromDelegatedAccount(permission);

  const keys: AccountMeta[] = [
    { pubkey: accounts.payer, isWritable: true, isSigner: true },
    { pubkey: ephemeralAta, isWritable: true, isSigner: false },
    { pubkey: PERMISSION_PROGRAM_ID, isWritable: false, isSigner: false },
    { pubkey: permission, isWritable: true, isSigner: false },
    { pubkey: SystemProgram.programId, isWritable: false, isSigner: false },
    { pubkey: buffer, isWritable: true, isSigner: false },
    { pubkey: record, isWritable: true, isSigner: false },
    { pubkey: metadata, isWritable: true, isSigner: false },
    { pubkey: DELEGATION_PROGRAM_ID, isWritable: false, isSigner: false },
    { pubkey: accounts.validator, isWritable: false, isSigner: false },
  ];

  const dataBuffer = Buffer.alloc(2);
  dataBuffer[0] = 7; // discriminator
  dataBuffer[1] = bump;

  return new TransactionInstruction({
    programId: EATA_PROGRAM_ID,
    keys,
    data: dataBuffer,
  });
}

File: rust/sdk/src/spl/builders/delegate_ephemeral_ata_permission.rs
use crate::pda::{
    delegate_buffer_pda_from_delegated_account_and_owner_program,
    delegation_metadata_pda_from_delegated_account, delegation_record_pda_from_delegated_account,
};

use crate::{
    access_control::structs::Permission,
    compat,
    consts::{ESPL_TOKEN_PROGRAM_ID, PERMISSION_PROGRAM_ID},
    cpi::DELEGATION_PROGRAM_ID,
    spl::{EphemeralAta, EphemeralSplDiscriminator},
};

pub struct DelegateEphemeralAtaPermissionBuilder {
    pub payer: compat::Pubkey,
    pub user: compat::Pubkey,
    pub mint: compat::Pubkey,
    pub validator: compat::Pubkey,
}

impl DelegateEphemeralAtaPermissionBuilder {
    #[inline(always)]
    pub fn instruction(&self) -> compat::Instruction {
        let (eata, _eata_bump) = EphemeralAta::find_pda(&self.user, &self.mint);
        let (permission, _permission_bump) = Permission::find_pda(&eata);
        let delegation_buffer = delegate_buffer_pda_from_delegated_account_and_owner_program(
            &permission,
            &PERMISSION_PROGRAM_ID,
        );
        let delegation_record = delegation_record_pda_from_delegated_account(&permission);
        let delegation_metadata = delegation_metadata_pda_from_delegated_account(&permission);

        compat::Instruction {
            program_id: ESPL_TOKEN_PROGRAM_ID,
            accounts: vec![
                compat::AccountMeta::new(self.payer, true),
                compat::AccountMeta::new(eata, false),
                compat::AccountMeta::new_readonly(PERMISSION_PROGRAM_ID, false),
                compat::AccountMeta::new(permission, false),
                compat::AccountMeta::new_readonly(
                    compat::Pubkey::new_from_array(solana_system_interface::program::ID.to_bytes()),
                    false,
                ),
                compat::AccountMeta::new(delegation_buffer, false),
                compat::AccountMeta::new(delegation_record, false),
                compat::AccountMeta::new(delegation_metadata, false),
                compat::AccountMeta::new_readonly(DELEGATION_PROGRAM_ID, false),
                compat::AccountMeta::new_readonly(self.validator, false),
            ],
            data: vec![EphemeralSplDiscriminator::DelegateEphemeralAtaPermission as u8],
        }
    }
}
